### PR TITLE
standardize info diagram definitions

### PR DESCRIPTION
**Pull request by Yokozuna59**
## :bookmark_tabs: Summary

Currently, most diagrams aren't fully typed, which causes some expected behaviors.

Originally, I was planning to do that here in #4450 while migrating to the new parser, but it appears it's going to take some time since I'm not familiar with the repo structure and [the current DBs](https://github.com/mermaid-js/mermaid/pull/4450#issuecomment-1583042912) and still can't figure out a way to run the parser without the language server. I guess by adding and documenting the types and functions, I would be knowledgeable about the mermaid structure and each diagram style, definition, and DB, and then I would be able to use those DBs in the new parser.

Standardize diagrams definitions by:

- Converting JS code into TS.
- Remove `@ts-*` comments
- Add and use defined types and interfaces.
- Add more JSDoc comments.
- Standardize Diagrams Definition:

  ```mermaid
  ---
  title: Diagram Definition
  ---
  flowchart LR
  diagram --> diagramDb.ts
  diagram --> diagramDetector.ts
  diagram --> diagramRenderer.ts
  diagram --> diagramDiagram.ts
  diagram --> diagramTypes.ts
  diagram --> diagramStyles.ts
  ```

## :straight_ruler: Design Decisions

Describe the way your implementation works or what design decisions you made if applicable.

Diagram Definition:

- DB: similar to the existing one but with types and standardizing definitions.
- Detector: similar to the existing one but with types if missing.
- Renderer: similar to the existing one but with `svgDraw.js` if exists.
- Diagram: contains the diagram definition with db, parser, styles, etc.
- Types: contains diagram types and interfaces for DB, styles options, etc. i.g., [classTypes.ts](https://github.com/mermaid-js/mermaid/blob/75502d076e843d3eb932c86027849347df737e08/packages/mermaid/src/diagrams/class/classTypes.ts).
- Styles (*NOT SURE*): Either combining it with `Renderer` like `quadrant` chart or converting it to TS like `flowchart`.

### :clipboard: Tasks

Make sure you

- [x] :book: have read the [contribution guidelines](https://github.com/mermaid-js/mermaid/blob/develop/CONTRIBUTING.md)
- [x] :computer: have added unit/e2e tests (if appropriate)
- [ ] :notebook: have added documentation (if appropriate)
- [x] :bookmark: targeted `develop` branch

### CODE REVIEW

**Review by sidharthv96 — COMMENTED** [review]
This is excellent work, long overdue!
Keep going 🚀

File: packages/mermaid/src/diagrams/er/erDiagram.ts
Comment (Yokozuna59) [reply packages/mermaid/src/diagrams/er/erDiagram.ts:1]: Oh, just a typo, I’ll fix it shortly.

File: .vite/jisonTransformer.ts
Comment (Yokozuna59) [reply .vite/jisonTransformer.ts:1]: It appears it doesn’t, and yes, Old Jison doesn’t export types. I tried removing all TS comments to see which didn’t throw an error, and I removed them. And by adding more types, they will decrease rabidly.

File: packages/mermaid/src/diagrams/info/infoDiagram.ts
Comment (Yokozuna59) [reply packages/mermaid/src/diagrams/info/infoDiagram.ts:11]: Well do.

Do you have anything on your mind about strict styling definitions? I’m sure that the current approach of writing CSS code as strings with no type safety or highlighting or etc.

I came across [csstypes](https://github.com/frenic/csstype), not sure about bundling size or limitations. Do you have something in mind? I’m almost done with pie chart files, but I'm not sure what to do with the `styles.js` file yet.

File: packages/mermaid/src/diagrams/er/erDiagram.ts
Comment (Yokozuna59) [reply packages/mermaid/src/diagrams/er/erDiagram.ts]: Done.

File: packages/mermaid/src/diagrams/info/infoDiagram.ts
Comment (Yokozuna59) [reply packages/mermaid/src/diagrams/info/infoDiagram.ts]: And since I'm already changing the pie renderer, I might resolve this issue #3987 with it since it's just a mathematical error, not a svg side, but I'm not sure what is the best approach yet. Adding decimals to the chart would lead to more problems and break almost all existing Cypress tests.

Do you something in your mind? Or should I just leave and focus on converting files and adding types without resolving existing error?

Comment (sidharthv96) [reply packages/mermaid/src/diagrams/info/infoDiagram.ts]: > Do you something in your mind? Or should I just leave and focus on converting files and adding types without resolving existing error?

Let's keep this PR narrowly scoped to just fixing types. It might be the best to leave https://github.com/mermaid-js/mermaid/issues/3987 alone. Or we'd have to add a config for decimals.

> I came across [csstypes](https://github.com/frenic/csstype) and [vanilla-extract](https://github.com/vanilla-extract-css/vanilla-extract), not sure about bundling size or limitations. Do you have something in mind? I’m almost done with pie chart files, but I'm not sure what to do with the styles.js file yet.

I think [vanilla-extract](https://github.com/vanilla-extract-css/vanilla-extract) looks great, and we can use that as it won't be bundled because it's a build time dependency. 

But this also should be a PR of it's own as we could discuss in depth on the implications there.

File: packages/mermaid/src/diagrams/info/infoTypes.ts
Comment (sidharthv96) [inline packages/mermaid/src/diagrams/info/infoTypes.ts:6]: nit: Casing.

```suggestion
export interface InfoDB extends DiagramDB {
```

File: packages/mermaid/src/diagrams/info/infoTypes.ts
Comment (sidharthv96) [inline packages/mermaid/src/diagrams/info/infoTypes.ts:5]: You can remove comments like these that don't really explain anything, good names are enough. 
`Readable Names >>> Comments`.

File: packages/mermaid/src/diagram-api/types.ts
Comment (sidharthv96) [inline packages/mermaid/src/diagram-api/types.ts:30]: Make sure this change is propagated to places where styles is consumed.

Like [addStylesForDiagram(id, diagram.styles);](https://github.com/mermaid-js/mermaid/blob/af13f3430aa994e7dd258aa5964894c476dc00ac/packages/mermaid/src/diagram-api/diagramAPI.ts#L53-L54)

File: packages/mermaid/src/diagrams/info/infoDiagram.ts
Comment (Yokozuna59) [reply packages/mermaid/src/diagrams/info/infoDiagram.ts]: > Let's keep this PR narrowly scoped to just fixing types. It might be the best to leave #3987 alone. Or we'd have to add a config for decimals.

Okey, I'll be focusing more on converting files and adding types with JSDoc comments.

> I think [vanilla-extract](https://github.com/vanilla-extract-css/vanilla-extract) looks great, and we can use that as it won't be bundled because it's a build time dependency.
> 
> But this also should be a PR of it's own as we could discuss in depth on the implications there.

Sound great!

File: packages/mermaid/src/diagrams/info/infoTypes.ts
Comment (Yokozuna59) [reply packages/mermaid/src/diagrams/info/infoTypes.ts:6]: I took it from `DiagramDb` interface, and I thought of the same thing, `DB` should be in caps, but since it was like that I followed it, but I'll fix soon.

File: packages/mermaid/src/diagrams/info/infoTypes.ts
Comment (Yokozuna59) [reply packages/mermaid/src/diagrams/info/infoTypes.ts:5]: I left it there as a placeholder; I wasn't sure what to add more, so I left it like that, and yes, readable names are more than enough, but I guess we need to add some to DB functions even if the name was clear since they're kind of fuzzy, especially for those that have more than one or even object parameters.

File: packages/mermaid/src/diagram-api/types.ts
Comment (Yokozuna59) [reply packages/mermaid/src/diagram-api/types.ts:30]: Oh, I updated to see If it causes any issues but there was none so didn't pay much attention to that, I'll look into details soon.

File: packages/mermaid/src/diagrams/info/infoTypes.ts
Comment (sidharthv96) [reply packages/mermaid/src/diagrams/info/infoTypes.ts:6]: Yes, following convention was the right call for normal PRs. 👍 

But, here, you're fixing the whole section, so let's change to `DiagramDB` too.

File: packages/mermaid/src/diagram-api/types.ts
Comment (sidharthv96) [reply packages/mermaid/src/diagram-api/types.ts:30]: Try doing a search for `.styles`, there might be some useage in untyped files.

File: packages/mermaid/src/diagrams/info/infoTypes.ts
Comment (Yokozuna59) [reply packages/mermaid/src/diagrams/info/infoTypes.ts:6]: Great! Should I rename files to `diagramDB.ts`? or `diagramDb.ts` is fine?

Comment (sidharthv96) [reply packages/mermaid/src/diagrams/info/infoTypes.ts]: Renaming files with only case changes is sadly a problem with git.
Different people having different git configs and OSes on handling casing leads to errors. 
Leaving the file name as-is shold be fine.

Comment (Yokozuna59) [reply packages/mermaid/src/diagrams/info/infoTypes.ts]: Git marks files as [`Renamed`](https://github.com/mermaid-js/mermaid/pull/4486/files#diff-40265b49334bcbafb58c9046a0dfb71c5b05aafffcc19f3af4c32fc4c7c61661) if is has small partial difference, but not sure about the percentage exactly.

And since most DBs are in js, adding types and organize it will most likely make git mark those files as deleted and new files has been created as ts. So I'm not it would matter since most of them would be marked as deleted, but if you say so.

**Comment by Yokozuna59** [comment]
For `d3` type issues, it appears that `@types/d3` package isn't enough or might even have no use; we need to import the types for the specific used package, so I'll add more of those types when needed. I'm currently working on converting pie chart files, and the renderer needs `@types/d3-shape`, so I'll add it and try to resolve those type errors and remove `ts-*` comments.

File: packages/mermaid/src/diagrams/info/infoRenderer.ts
Comment (sidharthv96) [reply packages/mermaid/src/diagrams/info/infoRenderer.ts:29]: Don't worry about deprecations. We can address them afterwards. Now, atleast we know it's deprecated.

Comment (Yokozuna59) [reply packages/mermaid/src/diagrams/info/infoRenderer.ts:29]: never mind, wrong interface.

**Review by Yokozuna59 — COMMENTED** [review]
I'm not sure how to assign `Document` d3 type to the new `HTML` one, there are some differences, so I can't make it a union type, it will affect the regular `select('body')`.

If you've got a better design for this mess, let me know so I can follow it with other renders.

**Comment by Yokozuna59** [comment]
And while working on pie files, it appears that [`width`](https://github.com/mermaid-js/mermaid/tree/56a9625d955c2ba0c3b25f19a4692c6860a2704e/packages/mermaid/src/diagrams/pie/pieRenderer.js#L42-L44) has default, shouldn't be that setted inside [`PieDiagramConfig`](https://github.com/mermaid-js/mermaid/tree/56a9625d955c2ba0c3b25f19a4692c6860a2704e/packages/mermaid/src/config.type.ts#L226-L228)?

**Comment by sidharthv96** [comment]
> I'm not sure how to assign `Document` d3 type to the new `HTML` one, there are some differences, so I can't make it a union type, it will affect the regular `select('body')`.
> 
> If you've got a better design for this mess, let me know so I can follow it with other renders.


I tried some approaches, but none are staightforward. Cast it or use `@ts-ignore` for now.

**Comment by sidharthv96** [comment]
> And while working on pie files, it appears that [`width`](https://github.com/mermaid-js/mermaid/tree/56a9625d955c2ba0c3b25f19a4692c6860a2704e/packages/mermaid/src/diagrams/pie/pieRenderer.js#L42-L44) has default; shouldn't that be setted inside [`PieDiagramConfig`](https://github.com/mermaid-js/mermaid/tree/56a9625d955c2ba0c3b25f19a4692c6860a2704e/packages/mermaid/src/config.type.ts#L226-L228)? Or create a new function using that interface in the DB file and set all the default values there, and maybe use `clear` function (or maybe `resetConfig` since `clear` is more like clearing DB) for resetting configs back to default?
> 
> I guess the second approach is better, so it would have more consistent and inline DB definitions.

Seems like `useWidth` is expected inside config?

But, I feel we should do these in different PRs.  One PR with a ton of change is hard for the reviewers and usually might contain errors that's overlooked. 

Can you wrap up the changes that's done until now, so we can review and merge this PR?
You can continue fixing rest of the types in a different PR, so it's easier to review for other reviewers who might not have been following this from the beginning.

**Comment by Yokozuna59** [comment]
> Seems like `useWidth` is expected inside config?

In original config it's mark as optional, which make sense when the user want just to override one field in the diagram config, but it when rendering it doesn't make sense to check if the default isn't undefined. 

> But, I feel we should do these in different PRs. One PR with a ton of change is hard for the reviewers and usually might contain errors that's overlooked.

Makes sense; we haven't even converted other diagrams, and we already have 46 files changed. Let's leave this PR for an info diagram and some common stuff; what do you think?

> Can you wrap up the changes that's done until now, so we can review and merge this PR? You can continue fixing rest of the types in a different PR, so it's easier to review for other reviewers who might not have been following this from the beginning.

Sure, they can be listed as the following:

## Common

- Remove unnecessary `@ts-*` comments.
- Rename `DiagramDb` interface into `DiagramDB`.
- Set `styles` as optional in `DiagramDefinition`.
- Change `.match` into `.test` in diagram detectors.
- Use `DiagramLoader` interface in `loader` function in diagram detectors.
- Add `@types/d3-selection` packages to mermmaid dev dependencies.
- Add `DrawDefinition`, `ParseDirectiveDefinition`, `HTML`, and `SVG` types.

## Info Diagram

- Convert all files from JS into TS.
- Add types and use types.
- Remove empty `styles.js`.
- Organize and one more test case in `info.spec.ts`
- Remove unparsed stuff inside DB.
- create `InfoDB`, and `InfoFields` interfaces and use them.
- create `DEFAULT_INFO_DB` to be used for `.clear()` method.
- create new `clear` because common `clean` has no use.

File: packages/mermaid/src/diagrams/info/infoDb.ts
Comment (sidharthv96) [inline packages/mermaid/src/diagrams/info/infoDb.ts]: Why do we need this?

File: packages/mermaid/src/diagrams/info/infoDetector.ts
Comment (sidharthv96) [inline packages/mermaid/src/diagrams/info/infoDetector.ts:22]:
```ts
{ id, diagram }
{ id: id, diagram: diagram }
```

Both are equivalent. The first one is a concise syntax.

File: packages/mermaid/src/diagrams/info/infoDb.ts
Comment (Yokozuna59) [reply packages/mermaid/src/diagrams/info/infoDb.ts]: No, but since it's it's exists on the info grammar I left them there, I tried adding it and removing it but no effect on rendered diagram.

File: packages/mermaid/src/diagrams/info/infoDetector.ts
Comment (sidharthv96) [reply packages/mermaid/src/diagrams/info/infoDetector.ts]: Use the shorter syntax everywhere (that you changed). Let's leave the rest of instances as-is for now.

File: packages/mermaid/src/diagrams/info/infoTypes.ts
Comment (sidharthv96) [inline packages/mermaid/src/diagrams/info/infoTypes.ts:9]: This is a value, it shouldn't be in types. 
Let's move it to infoDb.ts

File: packages/mermaid/src/diagrams/info/infoDb.ts
Comment (sidharthv96) [inline packages/mermaid/src/diagrams/info/infoDb.ts:19]:
```suggestion
const db: InfoDB = {
  clear,
  setInfo,
  getInfo,
};
```

Other places too.

**Review by sidharthv96 — CHANGES_REQUESTED** [review]
Nice work!

File: packages/mermaid/src/diagrams/info/infoDetector.ts
Comment (Yokozuna59) [reply packages/mermaid/src/diagrams/info/infoDetector.ts]: I'll fix them now.

File: packages/mermaid/src/diagrams/info/infoTypes.ts
Comment (Yokozuna59) [reply packages/mermaid/src/diagrams/info/infoTypes.ts:9]: I thought the same think, but I said it would be cleaner since it not modifiable constant. I'll move it now.

File: packages/mermaid/src/diagrams/info/infoRenderer.ts
Comment (Yokozuna59) [reply packages/mermaid/src/diagrams/info/infoRenderer.ts]: fixed.

File: packages/mermaid/src/diagrams/info/infoDb.ts
Comment (Yokozuna59) [reply packages/mermaid/src/diagrams/info/infoDb.ts]: fixed.

File: packages/mermaid/src/diagrams/info/infoTypes.ts
Comment (Yokozuna59) [reply packages/mermaid/src/diagrams/info/infoTypes.ts]: fixed.

**Review by sidharthv96 — APPROVED** [review]

File: packages/mermaid/src/diagrams/info/infoDb.ts
Comment (Yokozuna59) [reply packages/mermaid/src/diagrams/info/infoDb.ts]: I fixed export default db, should I change others to? the `infoDiagram.ts` would be like this:

```ts
import { draw } from './infoRenderer.js';

export const diagram: DiagramDefinition = {
  ...
  renderer: draw,
};
```

Comment (sidharthv96) [reply packages/mermaid/src/diagrams/info/infoDb.ts]: Not really. 
`export default { draw };` is the export in `infoRenderer`. 
So, the change would be
```ts
// infoRenderer.ts
export const renderer = { draw };
```

```ts
// infoDiagram.ts
import { renderer } from './infoRenderer.js';
export const diagram: DiagramDefinition = {
  ...
  renderer,
};
```

**Comment by Yokozuna59** [comment]
@sidharthv96 please re-review latest changes so I can uses the new types in pie chart files.

File: packages/mermaid/src/diagrams/info/infoDetector.ts
Comment (sidharthv96) [inline packages/mermaid/src/diagrams/info/infoDetector.ts:24]:
```ts
export const info: ExternalDiagramDefinition = {
  id,
  detector,
  loader,
};
```

I think this should be fine.

Comment (Yokozuna59) [reply packages/mermaid/src/diagrams/info/infoDetector.ts]: Done.

File: cypress/platform/info.html
Comment (Yokozuna59) [reply cypress/platform/info.html:1]: Removed.

**Comment by Yokozuna59** [comment]
Should I mock `infoRender` inside [`__mocks__`](https://github.com/mermaid-js/mermaid/tree/develop/__mocks__) dir? I know it has no use but just to standardize definitions.

And I'm currently working on pie chart files, should try supporting [`@vanilla-extract/css`](https://github.com/vanilla-extract-css/vanilla-extract) with it? Or just focus on converting files?

**Comment by sidharthv96** [comment]
> Should we mock `infoRenderer` inside [`__mocks__`](https://github.com/mermaid-js/mermaid/tree/develop/__mocks__) dir? 

You can add it if it's required to run any tests. No need to add and maintain files just for standardizing. 
---

> And add `cy.get('svg');` to `cypress/integration/rendering/info.spec.ts` after each test case?

You can actually remove all the `cy.get('svg')` inside tests that call `imgSnapshotTest`. 
`cy.get('svg')` is already called inside `imgSnapshotTest`, so the direct call inside the tests are redundant. 
---

> And I'm currently working on pie chart files, should try supporting [`@vanilla-extract/css`](https://github.com/vanilla-extract-css/vanilla-extract) with it? Or just focus on converting files?

It's better not to mix things. I actually had that habit a while back (still do for small things 😅), but it's best to keep things separate.

**Comment by Yokozuna59** [comment]
> You can add it if it's required to run any tests. No need to add and maintain files just for standardizing.

Ok.

> You can actually remove all the `cy.get('svg')` inside tests that call `imgSnapshotTest`. `cy.get('svg')` is already called inside `imgSnapshotTest`, so the direct call inside the tests are redundant.

It would affect many files and be hard to review. I'll remove them when converting the test file of the diagram.

> It's better not to mix things. I actually had that habit a while back (still do for small things 😅), but it's best to keep things separate.

It will be the same thing at the end; when trying to add styles, it will affect and change many files, and then it would be better to make each PR a separate diagram. But ok, I'll leave it to separate PR.

**Pull request merged**
